app/src/src_gplay/org/breezyweather/sources/android/AndroidLocationSource.kt

```diff
@@ -114,67 +114,80 @@ open class AndroidLocationSource @Inject constructor() : LocationSource, Locatio
         return rxObservable {
-            cancel()
+            gpsLocation = null
+            gmsLocation = null
+            clearLocationUpdates()
 
             locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager?
-            fusedLocationClient = if (isGMSEnabled(context)) {
-                LocationServices.getFusedLocationProviderClient(context)
-            } else null
+            currentProvider = locationManager?.let { getBestProvider(it) } ?: currentProvider
+
             if (locationManager == null
                 || !hasPermissions(context)
                 || !isLocationEnabled(locationManager!!)
-                || getBestProvider(locationManager!!).also { currentProvider = it }
-                    .isEmpty()
+                || currentProvider.isEmpty()
             ) {
                 LogHelper.log(msg = "Location manager not ready, no permissions, no location enabled or no provider available")
                 throw LocationException()
             }
 
-            lastKnownLocation = getLastKnownLocation(locationManager!!)
-
+            // request constant gps updates
             locationManager!!.requestLocationUpdates(
                 currentProvider,
                 0L,
                 0F,
                 this@AndroidLocationSource,
                 Looper.getMainLooper()
             )
-            fusedLocationClient?.let { client ->
-                client.requestLocationUpdates(
-                    LocationRequest
-                        .create()
-                        .setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY)
-                        .setNumUpdates(1),
-                    gmsLocationCallback,
-                    Looper.getMainLooper()
-                )
-                client.lastLocation.addOnSuccessListener {
-                    gmsLastKnownLocation = it
+
+            // if GMS is enabled, try and get one location update from gms
+            if (isGMSEnabled(context)) {
+                fusedLocationClient = LocationServices.getFusedLocationProviderClient(context)
+
+                fusedLocationClient?.let { client ->
+                    val locationRequest = LocationRequest.Builder(
+                        Priority.PRIORITY_BALANCED_POWER_ACCURACY,
+                        1
+                    ).setMaxUpdates(1).build()
+
+                    client.requestLocationUpdates(
+                        locationRequest,
+                        gmsLocationCallback,
+                        Looper.getMainLooper()
+                    )
+
+                    client.lastLocation.addOnSuccessListener {
+                        gmsLocation = it
+                    }
                 }
             }
 
-            delay(TIMEOUT_MILLIS)
+            // wait X seconds for callbacks to set locations
+            for (i in 1..TIMEOUT_MILLIS / 1000) {
+                delay(1000)
```

Comment:
Don’t do that. You can use a Looper Handler and call `timer.removeCallbacksAndMessages(null)` once you’re done to get out of the loop.